### PR TITLE
FSET-1351: Add ability to import event schedules

### PR DESCRIPTION
**Pull request by fayimora**

### CODE REVIEW

File: app/model/persisted/assessmentcentre/Event.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package model.persisted.assessmentcentre
+
+import repositories.{ BSONDateTimeHandler, BSONLocalDateHandler, BSONMapHandler, BSONLocalTimeHandler }
+import org.joda.time.{ LocalDate, LocalTime }
+import play.api.libs.json.Json
+import reactivemongo.bson.Macros
+
+/**
```
Comment (benjstephenson) [inline app/model/persisted/assessmentcentre/Event.scala]: pedantic comment, but can you update your Intellij settings to remove this file header?

File: app/repositories/package.scala [hunk 2]
```diff
@@ -119,7 +121,9 @@ package object repositories {
 
     applicationAssessmentScoresRepository.collection.indexesManager.create(Index(Seq(("applicationId", Ascending)), unique = true)),
 
-    assessorAvailabilityRepository.collection.indexesManager.create(Index(Seq(("userId", Ascending)), unique = true))
+    assessorAvailabilityRepository.collection.indexesManager.create(Index(Seq(("userId", Ascending)), unique = true)),
+
+    assessmentEventsRepository.collection.indexesManager.create(Index(Seq(("eventType", Ascending), ("date", Ascending)), unique = false))
```
Comment (benjstephenson) [inline app/repositories/package.scala]: We will want to do queries on location and venue in the future, might be a good idea to add these indices too

File: app/model/persisted/assessmentcentre/Event.scala (hunk 1 above)
Comment (fayimora) [inline app/model/persisted/assessmentcentre/Event.scala]: Sure 👍

File: test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    "successfully saves and loads the file contents" in new GoodTestFixture {
+      val events: Seq[Event] = service.processCentres().futureValue
+
+      events.size mustBe 2
+      events.head.eventType mustBe "FSAC"
+      events(1).skillRequirements.getOrElse("QUALITY_ASSURANCE_COORDINATOR", "--") mustBe 0
+    }
+
+    "fails gracefully when any field is malformed" in new MalformedTestFixture {
+      an[Exception] must be thrownBy service.processCentres().futureValue
```
Comment (benjstephenson) [inline test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala]: What kind of exception are we expecting?

File: app/repositories/assessmentcentre/AssessmentEventsRepository.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+
+class AssessmentEventsMongoRepository(implicit mongo: () => DB)
+  extends ReactiveRepository[Event, BSONObjectID](CollectionNames.ASSESSMENT_EVENTS,
+    mongo, Event.eventFormat, ReactiveMongoFormats.objectIdFormats)
+  with AssessmentEventsRepository {
+
+  override def save(event: Event): Future[Unit] = {
+    collection.insert(event).map(_ => ())
```
Comment (benjstephenson) [inline app/repositories/assessmentcentre/AssessmentEventsRepository.scala]: We can work on this after we've finished the controller changes, but this needs to take a list of events so we can persist the whole collection in one go rather than one at a time which would create some error handling headaches.

File: test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala (hunk 3 above)
Comment (fayimora) [inline test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala]: @benjstephenson At the moment, I'm simply returning `Failure[Exception]`. I got rid of the multiple cases for each type of exception

Comment (benjstephenson) [inline test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala]: True facts :+1:

File: app/repositories/assessmentcentre/AssessmentEventsRepository.scala (hunk 4 above)
Comment (fayimora) [inline app/repositories/assessmentcentre/AssessmentEventsRepository.scala]: Yes I agree. I'll change this to take a list of events and then `bulkInsert` later.

File: app/controllers/AssessmentEventsController.scala [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+trait AssessmentEventsController extends BaseController {
+  val assessmentEventsRepository: AssessmentEventsRepository
+  val assessmentCenterParsingService: AssessmentCentreParsingService
+
+  def uploadAssessmentEvents() = Action.async { implicit request =>
+    assessmentCenterParsingService.processCentres().flatMap{ events =>
+      Logger.debug("Events have been processed!")
+      assessmentEventsRepository.save(events)
+    }.map { _ => Created }.recover { case _ => UnprocessableEntity }
```
Comment (benjstephenson) [inline app/controllers/AssessmentEventsController.scala]: I think this map should be on the result of the repository save, else you'll get an orphaned future

File: test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala (hunk 3 above)
Comment (fayimora) [inline test/services/assessmentcentre/AssessmentCentreParsingServiceSpec.scala]: @benjstephenson resolved?

**Comment by benjstephenson** [comment]
LGTM :+1:

File: app/model/persisted/assessmentcentre/SkillType.scala [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import reactivemongo.bson.{ BSON, BSONHandler, BSONString }
+
+import scala.language.implicitConversions
+
+object SkillType extends Enumeration {
+  type SkillType = Value
+
+  val ASSESSOR, DEPARTMENTAL_ASSESSOR, CHAIR, EXERCISE_MARKER, QUALITY_ASSURANCE_COORDINATOR, SIFTER = Value
+
+  implicit def toString(SkillType: SkillType): String = SkillType.toString
```
Comment (kjharkin) [inline app/model/persisted/assessmentcentre/SkillType.scala]: Why is this necessary and where is it used?

Comment (benjstephenson) [inline app/model/persisted/assessmentcentre/SkillType.scala]: the implicit toString can be useful when building BSON queries so we can avoid calling enumVal.toString.  We did this with the ApplicationStatuses enum too

Comment (kjharkin) [inline app/model/persisted/assessmentcentre/SkillType.scala]: Is it actually used though and if so where?
